src/style-spec/function/definitions/match.js

```diff
@@ -95,26 +95,30 @@ class Match implements Expression {
         return new Match(context.key, (inputType: any), (outputType: any), input, cases, outputs, otherwise);
     }
 
-    compile() {
-        const input = this.input.compile();
-        const outputs = [`function () { return ${this.otherwise.compile()} }`];
-        const lookup = {};
+    compile(ctx: CompilationContext) {
+        const input = ctx.compile(this.input);
 
-        for (const label in this.cases) {
-            // shift the index stored in this.cases by one, as we're using
-            // outputs[0] for the 'otherwise' case.
-            lookup[`${String(label)}`] = this.cases[label] + 1;
-        }
+        const outputs = [];
         for (const output of this.outputs) {
-            outputs.push(`function () { return ${output.compile()} }`);
+            outputs.push(ctx.addExpression(output.compile(ctx)));
         }
 
-        return `(function () {
-            var o = [${outputs.join(', ')}];
-            var l = ${JSON.stringify(lookup)};
-            var i = ${input};
-            return o[l[$this.as(i, ${JSON.stringify(this.inputType)})] || 0]();
-        })()`;
+        let lookup = '';
+        const labels = Object.keys(this.cases);
+        for (let i = 0; i < labels.length; i++) {
+            if (i > 0) lookup += ', ';
+            const label = labels[i];
+            lookup += `${String(label)}: ${outputs[this.cases[label]]}`;
+        }
+
+        const lookupObject = ctx.addVariable(`{${lookup}}`);
+
+        const inputType = this.inputType.kind !== 'Array' ?
+            `$this.types.${this.inputType.kind}` : JSON.stringify(this.inputType);
+
+        return `
+var i = $this.as(${input}, ${inputType});
+return (${lookupObject}[i] || ${ctx.addExpression(this.otherwise.compile(ctx))})();`;
```

Comment:
Maybe let's indent lines like this — we want to prioritize source code readability over generated one's.